src/audio/dai.c

```diff
@@ -688,7 +688,7 @@ static int dai_config(struct comp_dev *dev, struct sof_ipc_dai_config *config)
 	switch (config->type) {
```

Comment:
@plbossart the commit message looks very good. I think would be a nice addition to explain the current supported scenario and WHY do you need multi-dai configuration. The description must be made in natural language so that can be understood even by the users of SOF.

So, as far as I understand now once a topology is loaded we can support just 1 DAI with 1 configuration. For example, we can support playback over SSP for PCM streams with rate 48KHz, channels 2, format S16_LE.

Is that right?

Now, if we want to play a file with 4 channels for example this is not really possible due to limited topology?

Going forward: With this change would be somehow possible to specify a topology that supports multiple DAIs like for example on i.MX8 -> ESAI and SAI, and also support multiple configurations (like for example we can support playing files at 44.1Khz / 48Khz) dynamically without the need to reload the topology?

Is my understanding correct? If so, this is a great feature!

Reply:
The multi-dai is really an 'aggregation' feature, so that a stream can be rendered on two separate interfaces. For SoundWire, we have e.g. the left amplifier on one link and the right amplifier on the second. The multi-dai support provides you with the ability to build a 'compound' dai from existing components. From the kernel side you only see one DAI, but the firmware will program and synchronize the individual components. It's an alternative to the 'multi-cpu dai' approach at the kernel level, which could have been a solution as well but might be more complicated to handle at the firmware level.

The multi-configuration is unrelated, it's rather to support different modes of operation. We need this for Bluetooth support, where we can either use 48kHz 2ch 24 bits or 16kHz mono 16 bits. There is no dynamic switch between configs, and it may not be usable by e.g. PulseAudio, but a dedicated Bluetooth app or stack would know what to do.

I'll update the commit messages, for now I am in 'don't break anything' validation mode.

Reply:
@plbossart I see. I'm very interested about multi-configuration stuff. What do you understand by dynamic switching between configs?

Ideally, I'm wishing this scenario to work:

1. Run aplay -Dhw:0,0 -f S32_LE -c 2 -r 48000 sample.wav

At some point this will call hw_params() in Linux kernel which in turn will set the correct configuration on FW.

Then:

2. Run aplay -Dhw:0,0 -f S32_LE -c 4 -r 44100 sample.wav

Similarly, this will call hw_params() in Linux kernel which in turn will set the correct configuration on FW.

AFAIK this use case is not really supported because the topology on FW is created at the same time when the sound card is created on Linux kernel.

Reply:
that's exactly what @ranj063 and I were talking about earlier, we'd need the kernel to send the config IPC at hw_params time for all DAIs (currently it's only done for HDaudio and ALH). the kernel would select which config_index is desired and the firrmware woud apply configuration1 for 48 kHZ and configuration2 for 44.1kHz.

The point about switching is that you have to close the PCM and shutdown the DAI before you can change configurations. So you can still do run-time changes, but not after the hw_params were selected - if you do you have to go back to square1.

Reply:
@plbossart Ok, we are on the same page. I only need to switch configuration when PCM is opened and that's all. If i need another configuration I will close the PCM and re-open it with new params. 

This will work for us for PCM streams. We will see how things will be going with the compress interface where there I think there is a need to change the params without reopening the interface. But that's another story.

My concern now is about the initial configuration of the pipeline. Because on FW the pipeline is created just only once when the sound card is created on the Linux kernel side. When pipeline is created one needs to know how much the buffers will be, for example.

So, when sound card is created the kernel will look at PCM card capabilities and extract the maximum possible parameters.

If the sound card on the Linux kernel is advertised to support the following config:

rate: 44100, 48000
chanels: 2, 4, 16
rate: S16_LE, S24_LE, S32_LE

Then the kernel needs to figure out that, it needs to support:
maxrate: 48000
maxchan: 16
maxformat: S32_LE

and then create the pipeline on the firmware with exact these parameters.

Reply:
On the PCM/host side, you can have a lot more variations as long as you have the proper converter for the DAI format. For the DAIs you should really have 3 or 4 configs max.

Note that the pipeline creation on startup is what we do today, at some point we should really need to create the pipeline when it's needed. That would really allow for more optimal uses of the hardware resources.